### DISCUSSION TITLE
Xiaomi Sensors variable MQTT payload

### DISCUSSION THREAD

**Discussion by Kollejoe** [arendst/Tasmota#10942] [Q&A] [answered]
Hi Folks,

i have a small issue using Xiaomi Sensors with Sonoff ZbBridge and Tasmota on it.
Im using OpenHAB to administrate my Smart Home.

Now my problem is, that sometimes i don't receive the full payload. 
Here is a example payload for Xiaomi Door Sensor.

`{
  "ZbReceived": {
    "Fr_GsKlo": {
      "Device": "0xA8D1",
      "Name": "Fr_GsKlo",
      "Contact": 1,
      "Endpoint": 1,
      "LinkQuality": 92
    }
  }
}`

What im trying to read is the **Contact** value.
But then sometimes the payload looks like this:
`{
  "ZbReceived": {
    "Fr_GsKlo": {
      "Device": "0xA8D1",
      "Name": "Fr_GsKlo",
      "BatteryVoltage": 3,
      "BatteryPercentage": 98,
      "Endpoint": 1,
      "LinkQuality": 63
    }
  }
}`

Receiving this i get a lot of warnings like: 
`Executing the JSONPATH-transformation failed: Invalid path '$.ZbReceived.Fr_GsKlo.Contact' in '{"ZbReceived":{"Fr_GsKlo":{"Device":"0xA8D1","Name":"Fr_GsKlo","BatteryVoltage":3,"BatteryPercentage":98,"Endpoint":1,"LinkQuality":63}}}'`
coz there is no **Contact** value inside the payload.

I meat it still works, but i would like to get rid of these warning.

So unfortunately I was not able to find something on the net, maybe i was not thoroughly enough or I'm to stupid :(

Do you guys have a idea what could help here?
Do someone also use Sonoff ZbBridge with Xiaomi sensors and saw this behavior? 
Maybe there is a option to enlarge the payload, or something...

Many thanks
Niko

**Comment by sfromis** [comment, marked as answer]
Well, both payloads you post are "complete", they are just for different events, not unique to Xiaomi sensors. The one with battery data is just some additional info which may be useful for being ready to switch the battery. What I do with similar payloads is to check if the payload is of the "right type" before trying to parse fields which may or may not be there.

**Reply by Kollejoe** [reply]
First of all, thanks for that fast reply! 
And i've got your point! I agree with ascillato2 to mark your answer as a solution :)

Maybe one more question: is this sending of different events with different payloads something Tasmota specific? I mean who is responsible for what is there in the payload at the end?

**Reply by sfromis** [reply]
The way Zigbee2Tasmota works is that it tries to do a generalized transformation of the received data to more understandable JSON payloads, based on the original messages. I've not been involved in the considerations about how to structure that, but I can see that the result is that they are published to the same topic and general structure, with only the fields being different. Of course, it could have been done differently, with a more clear separation between types of messages, but with lots of different Zigbee devices available, the focus has been on "just" providing the data available.

I'm not aware of any "standard" for how to build JSON payloads, as this is more a way of sending data structures across a communication channel, with the result being whatever each creator came up with.

**Reply by Kollejoe** [reply]
Hey Buddy,

sorry for the delay. Was a bit busy...
Thanks for your reply! 👍 

Ya sure, i understand the pragmatic idea behind the actual implementation but its not that user friendly. Maybe there will be improvement in the future, who knows. :)

Anyway thanks one more time. I wish you success at your projects and stay healthy!

Cheers
